### ISSUE TITLE
Should there be a guide for people doing development on Npcap?

### ISSUE THREAD

**Issue by guyharris** [nmap/nmap#2003]
I'm trying to figure out how to profile a test program (it runs `pcap_findalldevs()` many times, to try to figure out where the CPU time is going), but apparently this may require 1) .pdb files for wpcap.dll and packet.dll and 2) might also require building the libraries with /PROFILE.

Should there be a guide for people doing development work on Npcap itself, indicating how to set up the development environment?  It might belong in the "Npcap internals" documentation.

**Comment by dmiller-nmap** [comment]
Yeah, that sounds like a good idea. So far all we have is [a section in the Npcap Guide](https://nmap.org/npcap/guide/npcap-devguide.html) that is far from complete.

Regarding profiling/debugging, we distribute PDBs of packet.dll and all our binaries in the Debug Symbols package on npcap.org, but we don't have wpcap.dll because I can't figure out how to get CMake to produce PDB files. Any hints/clues?

**Comment by guyharris** [comment]
> Regarding profiling/debugging, we distribute PDBs of packet.dll and all our binaries in the Debug Symbols package on npcap.org,

OK, I got those, and it got deeper into the call tree.

Now I'm burning up our Internet connection getting the symbols for Microsoft's libraries; hopefully that'll go all the way down to ntdll.dll.

> but we don't have wpcap.dll because I can't figure out how to get CMake to produce PDB files. Any hints/clues?

At least on the builds I'm doing on the command line with msbuild, on Windows 10 with VS 2019, it puts both pcap.dll and pcap.pdb into run\Debug.  Does that also work with RelWithDebInfo builds?

**Comment by guyharris** [comment]
> > Regarding profiling/debugging, we distribute PDBs of packet.dll and all our binaries in the Debug Symbols package on npcap.org,
> 
> OK, I got those, and it got deeper into the call tree.
> 
> Now I'm burning up our Internet connection getting the symbols for Microsoft's libraries; hopefully that'll go all the way down to ntdll.dll.

Yup - that works. It even looks as if it may dig into kernel-mode code, below ntdll.dll.  (I think I've seen that in some macOS sample stack traces as well.)

**Comment by dmiller-nmap** [comment]
Yes, that works, but then wpcap.dll is twice as big as the Release build, probably because of different `<InlineFunctionExpansion>` settings in the vcxproj. I think we'll post-process `wpcap.vcxproj` to change the `<GenerateDebugInformation>` value to `true`. It'd be nice if there was a CMake way to do this, though.

**Comment by dmiller-nmap** [comment]
We ended up making a `.props` file in our private build repo that sets compiler options `/Zi /Ob2` and linker options `/DEBUG:FULL /OPT:REF,ICF`. This creates the appropriate PDB file without increasing the size of the deployed DLL.

**Comment by dmiller-nmap** [comment]
wpcap.dll symbols are now distributed in the DebugSymbols package for Npcap 0.9991 and later. This issue is now about documentation for debugging/profiling Npcap.

**Comment by guyharris** [comment]
> This issue is now about documentation for debugging/profiling Npcap.

...including building it in the first place.
